File: renamer.py
import argparse
import os
import re
from tqdm import tqdm

def rename_files(directory, args, is_recursive=False, level=0):
    try:
        files = os.listdir(directory)
    except FileNotFoundError:
        print(f"指定されたディレクトリが存在しません: {directory}")
        os.makedirs(directory)
        print(f"ディレクトリを作成しました: {directory}")
        return
    except Exception as e:
        print(f"ディレクトリの読み込み中にエラーが発生しました: {e}")
        return

    file_count = len(files)
    num_length = len(str(file_count))

    for i, filename in enumerate(tqdm(files, desc="Processing"), start=1):
        old_path = os.path.join(directory, filename)
        if os.path.isdir(old_path) and not args.folder:
            if is_recursive:
                rename_files(old_path, args, is_recursive=True, level=level + 1)
            continue
        elif not os.path.isdir(old_path) and not args.file:
            continue

        if os.path.isdir(old_path):
            extension = ""
            base_name = filename
        else:
            base_name, extension = os.path.splitext(filename)
        
        new_name = modify_name(i, base_name, num_length, args) + extension

        if new_name != filename:
            new_path = os.path.join(directory, new_name)
            if args.debug:
                print(f"デバッグモード: '{filename}' から '{new_name}' への変更をシミュレートします。")
                continue  # 実際の名前変更をスキップ
            else:
                try:
                    os.rename(old_path, new_path)
                    print(f"'{filename}' を '{new_name}' に変更しました。")
                except Exception as e:
                    print(f"ファイル名の変更中にエラーが発生しました: {e}")

def modify_name(index, base_name, num_length, args):
    new_name = base_name

    if args.del_first:
        new_name = new_name[args.del_first:]

    if args.del_last:
        new_name = new_name[:-args.del_last]

    if args.add_first:
        new_name = args.add_first + new_name

    if args.add_last:
        new_name = new_name + args.add_last

    if args.add_number_first:
        new_name = f"{index:0{num_length}d}_" + new_name

    if args.add_number_last:
        new_name = new_name + f"_{index:0{num_length}d}"

    # --replaceオプションの処理を修正
    if args.replace:
        old_patterns = args.replace[:-1]
        new_str = args.replace[-1]
        for old_pattern in old_patterns:
            pattern = re.compile(re.escape(old_pattern))
            new_name = pattern.sub(new_str, new_name)

    if args.del_after:
        pos = new_name.find(args.del_after)
        if pos != -1:
            new_name = new_name[:pos + len(args.del_after)]

    if args.del_before:
        pos = new_name.find(args.del_before)
        if pos != -1:
            new_name = new_name[pos:]

    if args.add_after:
        search_str, add_str = args.add_after.split(",")
        pos = new_name.find(search_str)
        if pos != -1:
            new_name = new_name[:pos + len(search_str)] + add_str + new_name[pos + len(search_str):]

    if args.add_before:
        search_str, add_str = args.add_before.split(",")
        pos = new_name.find(search_str)
        if pos != -1:
            new_name = new_name[:pos] + add_str + new_name[pos:]

    if args.reg_del:
        pattern = re.compile(args.reg_del)
        new_name = pattern.sub('', new_name)

    if args.reg_del_around:
        pattern = re.compile(args.reg_del_around)
        match = pattern.search(new_name)
        if match:
            new_name = new_name[match.start():match.end()]
        else:
            new_name = base_name  # 修正: マッチしなかった場合の処理を明確化

    return new_name

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="ファイル名を一括で変更するスクリプト")
    parser.add_argument("--dir", type=str, help="対象ディレクトリ")
    parser.add_argument("--recursive", action="store_true", help="再帰的に処理を行う")
    parser.add_argument("--folder", action="store_true", help="フォルダを対象に含める")
    parser.add_argument("--file", action="store_true", help="ファイルを対象に含める")
    parser.add_argument("--del_first", type=int, help="先頭から削除する文字数")
    parser.add_argument("--del_last", type=int, help="末尾から削除する文字数")
    parser.add_argument("--add_first", type=str, help="先頭に追加する文字列")
    parser.add_argument("--add_last", type=str, help="末尾に追加する文字列")
    parser.add_argument("--add_number_first", action="store_true", help="先頭に番号を追加する")
    parser.add_argument("--add_number_last", action="store_true", help="末尾に番号を追加する")
    parser.add_argument("--replace", nargs='+', help="指定文字列を置換 (最後に定義された文字列を置換先として扱う) 例：--replace \"置換元A\" \"置換元B\" \"置換先\"")
    parser.add_argument("--del_after", type=str, help="指定文字列の後を削除")
    parser.add_argument("--del_before", type=str, help="指定文字列の前を削除")
    parser.add_argument("--add_after", type=str, help="指定文字列の後に追加 (カンマで区切る)")
    parser.add_argument("--add_before", type=str, help="指定文字列の前に追加 (カンマで区切る)")
    parser.add_argument("--reg_del", type=str, help="正規表現で一致した部分を削除")
    parser.add_argument("--reg_del_around", type=str, help="正規表現で一致した部分を残し、それ以外を削除")
    parser.add_argument("--debug", action="store_true", help="デバッグモード")

    args = parser.parse_args()

    if args.recursive:
        rename_files(args.dir, args, is_recursive=True)
    else:
        rename_files(args.dir, args)

File: JavaScript/renamer.js
const fs = require('fs');
const path = require('path');
const os = require('os');

// コマンドライン引数の解析
const args = parseArgs(process.argv.slice(2));

// デバッグモードの設定
const isDebugMode = args['--debug'];

// 処理対象ディレクトリの設定
const targetDir = args['--dir'] || '.';

// 出力ディレクトリの設定
const outputDir = args['--save_dir'] || 'output';

// 処理対象の拡張子の設定
const targetExtensions = args['--extension'] ? args['--extension'].split(',') : ['txt'];

// 再帰的な処理の設定
const isRecursive = args['--recursive'];

// 先頭文字削除数の設定
const delFirstCount = parseInt(args['--del_first']) || 0;

// 末尾文字削除数の設定
const delLastCount = parseInt(args['--del_last']) || 0;

// 先頭に追加する文字列の設定
const addFirst = args['--add_first'] || '';

// 末尾に追加する文字列の設定
const addLast = args['--add_last'] || '';

// 先頭に連番を追加するかの設定
const addNumberFirst = args['--add_number_first'];

// 末尾に連番を追加するかの設定
const addNumberLast = args['--add_number_last'];

// 置換する文字列の設定
const replaceFrom = args['--replace'] ? args['--replace'].split(' ')[0] : '';
const replaceTo = args['--replace'] ? args['--replace'].split(' ')[1] : '';

// 指定した文字以降を削除するかの設定
const delAfter = args['--del_after'];

// 指定した文字以前を削除するかの設定
const delBefore = args['--del_before'];

// 指定した文字の後ろに文字を追加するかの設定
const addAfter = args['--add_after'] ? args['--add_after'].split(',')[0] : '';
const addAfterString = args['--add_after'] ? args['--add_after'].split(',')[1] : '';

// 指定した文字の前に文字を追加するかの設定
const addBefore = args['--add_before'] ? args['--add_before'].split(',')[0] : '';
const addBeforeString = args['--add_before'] ? args['--add_before'].split(',')[1] : '';

// 正規表現でマッチした箇所を削除するかの設定
const regDelPattern = args['--reg_del'] ? new RegExp(args['--reg_del']) : null;

// 正規表現でマッチした箇所以外を削除するかの設定
const regDelAroundPattern = args['--reg_del_around'] ? new RegExp(args['--reg_del_around']) : null;

// フォルダ名のみを対象とするかの設定
const isFolderOnly = args['--folder'];

// ファイル名のみを対象とするかの設定
const isFileOnly = args['--file'];

// メモリキャッシュの設定
const useMemCache = args['--mem_cache'] !== 'OFF';

// マルチスレッド処理の設定
const threadCount = parseInt(args['--threads']) || Math.max(1, os.cpus().length - 1);

// 処理対象のファイルパスを取得する
const targetFiles = getTargetFiles(targetDir, targetExtensions, isRecursive);

// ファイル名の変更処理
renameFiles(targetFiles, outputDir, useMemCache, threadCount);

/**
 * コマンドライン引数を解析する
 * @param {string[]} args コマンドライン引数
 * @returns {Object} 解析された引数
 */
function parseArgs(args) {
  const parsedArgs = {};
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg.startsWith('--')) {
      const key = arg;
      let value = true;
      if (args[i + 1] && !args[i + 1].startsWith('--')) {
        value = args[i + 1];
        i++;
      }
      parsedArgs[key] = value;
    }
  }
  return parsedArgs;
}

/**
 * 処理対象のファイルパスを取得する
 * @param {string} dir 処理対象ディレクトリ
 * @param {string[]} extensions 処理対象の拡張子
 * @param {boolean} isRecursive 再帰的な処理を行うか
 * @returns {string[]} 処理対象のファイルパス
 */
function getTargetFiles(dir, extensions, isRecursive) {
  const targetFiles = [];
  const files = fs.readdirSync(dir);
  for (const file of files) {
    const filePath = path.join(dir, file);
    const stats = fs.statSync(filePath);
    if (stats.isDirectory()) {
      if (isRecursive) {
        targetFiles.push(...getTargetFiles(filePath, extensions, isRecursive));
      }
    } else {
      const ext = path.extname(file).slice(1);
      if (extensions.includes(ext)) {
        targetFiles.push(filePath);
      }
    }
  }
  return targetFiles;
}

/**
 * ファイル名を変更する
 * @param {string} fileName ファイル名
 * @returns {string} 変更後のファイル名
 */
function renameFile(fileName) {
  let newFileName = path.basename(fileName, path.extname(fileName));
  const ext = path.extname(fileName);

  // 先頭文字削除
  if (delFirstCount > 0) {
    newFileName = newFileName.slice(delFirstCount);
  }

  // 末尾文字削除
  if (delLastCount > 0) {
    newFileName = newFileName.slice(0, -delLastCount);
  }

  // 先頭に文字列を追加
  if (addFirst) {
    newFileName = addFirst + newFileName;
  }

  // 末尾に文字列を追加
  if (addLast) {
    newFileName += addLast;
  }

  // 先頭に連番を追加
  if (addNumberFirst) {
    newFileName = `${fileCounter++}_${newFileName}`;
  }

  // 末尾に連番を追加
  if (addNumberLast) {
    newFileName += `_${fileCounter++}`;
  }

  // 文字列を置換
  if (replaceFrom) {
    newFileName = newFileName.replace(new RegExp(replaceFrom, 'g'), replaceTo);
  }

  // 指定した文字以降を削除
  if (delAfter) {
    newFileName = newFileName.split(delAfter)[0];
  }

  // 指定した文字以前を削除
  if (delBefore) {
    newFileName = newFileName.split(delBefore).pop();
  }

  // 指定した文字の後ろに文字列を追加
  if (addAfter) {
    newFileName = newFileName.replace(new RegExp(addAfter, 'g'), `$&${addAfterString}`);
  }

  // 指定した文字の前に文字列を追加
  if (addBefore) {
    newFileName = newFileName.replace(new RegExp(addBefore, 'g'), `${addBeforeString}$&`);
  }

  // 正規表現でマッチした箇所を削除
  if (regDelPattern) {
    newFileName = newFileName.replace(regDelPattern, '');
  }

  // 正規表現でマッチした箇所以外を削除
  if (regDelAroundPattern) {
    newFileName = newFileName.match(regDelAroundPattern)[0];
  }

  // フォルダ名のみを対象とする場合
  if (isFolderOnly) {
    return path.join(path.dirname(fileName), newFileName);
  }

  // ファイル名のみを対象とする場合
  if (isFileOnly) {
    return path.join(path.dirname(fileName), `${newFileName}${ext}`);
  }

  // デフォルト
  return path.join(path.dirname(fileName), `${newFileName}${ext}`);
}

/**
 * ファイル名を変更する (マルチスレッド対応)
 * @param {string[]} targetFiles 処理対象のファイルパス
 * @param {string} outputDir 出力ディレクトリ
 * @param {boolean} useMemCache メモリキャッシュを使用するか
 * @param {number} threadCount スレッド数
 */
function renameFiles(targetFiles, outputDir, useMemCache, threadCount) {
  const totalFiles = targetFiles.length;
  let processedFiles = 0;
  let fileCounter = 1;

  // 出力ディレクトリが存在しない場合は作成する
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // メモリキャッシュを使用する場合
  if (useMemCache) {
    const memCache = {};

    // マルチスレッド処理
    const pool = new MemoryPool(threadCount);
    targetFiles.forEach((filePath) => {
      pool.run(() => {
        const newFilePath = renameFile(filePath);
        memCache[filePath] = newFilePath;
        processedFiles++;
        printProgress(processedFiles, totalFiles);
      });
    });
    pool.drain(() => {
      // メモリキャッシュから出力ディレクトリにファイルを書き込む
      for (const [oldPath, newPath] of Object.entries(memCache)) {
        const newDir = path.join(outputDir, path.relative(targetDir, path.dirname(newPath)));
        if (!fs.existsSync(newDir)) {
          fs.mkdirSync(newDir, { recursive: true });
        }
        const newFilePath = path.join(newDir, path.basename(newPath));
        fs.copyFileSync(oldPath, newFilePath);
      }
      console.log('処理が完了しました。');
    });
  } else {
    // メモリキャッシュを使用しない場合
    targetFiles.forEach((filePath) => {
      const newFilePath = renameFile(filePath);
      const newDir = path.join(outputDir, path.relative(targetDir, path.dirname(newFilePath)));
      if (!fs.existsSync(newDir)) {
        fs.mkdirSync(newDir, { recursive: true });
      }
      const newFullPath = path.join(newDir, path.basename(newFilePath));
      fs.copyFileSync(filePath, newFullPath);
      processedFiles++;
      printProgress(processedFiles, totalFiles);
    });
    console.log('処理が完了しました。');
  }
}

/**
 * 処理の進捗状況を表示する
 * @param {number} processed 処理済みのファイル数
 * @param {number} total 総ファイル数
 */
function printProgress(processed, total) {
  const progress = Math.floor((processed / total) * 100);
  const bar = '█'.repeat(Math.floor(progress / 5)) + '░'.repeat(20 - Math.floor(progress / 5));
  process.stdout.write(`\r[${bar}] ${progress}% (${processed}/${total})`);
  if (processed === total) {
    process.stdout.write('\n');
  }
}

/**
 * マルチスレッド処理用のプール
 */
class MemoryPool {
  constructor(threadCount) {
    this.threadCount = threadCount;
    this.queue = [];
    this.workers = [];
    this.createWorkers();
  }

  createWorkers() {
    for (let i = 0; i < this.threadCount; i++) {
      const worker = {
        thread: null,
        available: true,
      };
      this.workers.push(worker);
    }
  }

  run(task) {
    const availableWorker = this.workers.find((worker) => worker.available);
    if (availableWorker) {
      availableWorker.available = false;
      availableWorker.thread = new Promise((resolve) => {
        task();
        resolve();
      }).finally(() => {
        availableWorker.available = true;
        this.run(this.queue.shift());
      });
    } else {
      this.queue.push(task);
    }
  }

  drain(callback) {
    Promise.all(this.workers.map((worker) => worker.thread || Promise.resolve()))
      .then(callback)
      .catch((err) => console.error(err));
  }
}
